llap-ext-client/src/java/org/apache/hadoop/hive/llap/LlapBaseInputFormat.java

```diff
@@ -345,8 +346,11 @@ public static void close(String handleId) throws IOException {
    */
   public static void closeAll() {
     LOG.debug("Closing all handles");
-    for (String handleId : connectionMap.keySet()) {
+    Iterator<String> handleIds = connectionMap.keySet().iterator();
+    String handleId = null;
+    while (handleIds.hasNext()) {
       try {
+        handleId = handleIds.next();
```

Comment:
Looking at https://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html, esp. the following para

> The iterators returned by all of this class's "collection view methods" are fail-fast: if the map is structurally modified at any time after the iterator is created, in any way except through the iterator's own remove method, the iterator will throw a ConcurrentModificationException. Thus, in the face of concurrent modification, the iterator fails quickly and cleanly, rather than risking arbitrary, non-deterministic behavior at an undetermined time in the future.

it looks like even the new code is prone to the same exception. Is the synchronize block in close() protects the code from exception?. Ideally we should be using iterator.remove() method to remove the element while iterating over keySet().

I think, what we want to do here is
1. Write two close() methods, one accepting a handleId and other accepting handleConnections. The first gets the handleConnections from handleId and calls the second.
2. Modify closeAll() method to iterate over the connectionMap.keySet() using an interator. For every element in the interator, get the handleConnections and call close() method with those handleConnections. Now call remove() on the iterator.

Does that look good?

Reply:
Thanks Ashutosh, good catch. 
I just realized, even without doing all these changes we can avoid it by just changing the original code `for (String handleId : connectionMap.keySet())` 
to `for (String handleId : new ArrayList<>(connectionMap.keySet()))`

Looks good?

Reply:
I looked up many web pages/blogs which discuss "removing an element from HashMap while iterating over it". All of them suggest one common method i.e. using an iterator. I saw using ArrayList somewhere, but that's not a common suggestion. If the keySet() doesn't return all the keys at once to create the ArrayList at once, which it shouldn't ideally, we may be removing a key which the iterator is scanning, thus causing an exception.

Since at the end we will remove all the elements, one possible solution is to close all the connections first by iterating over the Map and then call clear() on the Map. This will still require us to refactor close() method as suggested earlier.

Reply:
Added another patch :)